File: csharp/Tracker.cs
using System.Collections.Concurrent;
using Royaltics.ErrorTracker.Types;

namespace Royaltics.ErrorTracker;

public static class Tracker
{
    private static readonly ConcurrentDictionary<string, ErrorTrackerClient> Instances = new();
    private static ErrorTrackerClient? _defaultInstance;
    private static readonly object Lock = new();

    public static ErrorTrackerClient Create(ClientConfig config, string? name = null)
    {
        var client = new ErrorTrackerClient(config);

        if (name != null)
        {
            Instances[name] = client;
        }
        else
        {
            lock (Lock)
            {
                _defaultInstance ??= client;
            }
        }

        return client.Start();
    }

    public static ErrorTrackerClient Get(string? name = null)
    {
        if (name != null)
        {
            if (!Instances.TryGetValue(name, out var instance))
            {
                throw new InvalidOperationException($"Tracker instance \"{name}\" not found");
            }
            return instance;
        }

        if (_defaultInstance == null)
        {
            throw new InvalidOperationException("No default tracker initialized. Call Tracker.Create() first.");
        }

        return _defaultInstance;
    }

    public static ErrorTrackerClient Error(
        Exception error,
        EventLevel level = EventLevel.ERROR,
        Dictionary<string, string>? metadata = null)
    {
        return Get().Error(error, level, metadata);
    }

    public static ErrorTrackerClient Event(
        string title,
        EventLevel level = EventLevel.INFO,
        Dictionary<string, string>? metadata = null)
    {
        return Get().Event(title, level, metadata);
    }

    public static Task FlushAsync()
    {
        return Get().ForceFlushAsync();
    }

    public static ErrorTrackerClient Pause()
    {
        return Get().Pause();
    }

    public static ErrorTrackerClient Resume()
    {
        return Get().Resume();
    }

    public static void Shutdown()
    {
        lock (Lock)
        {
            _defaultInstance?.Shutdown();
            _defaultInstance = null;
        }

        foreach (var instance in Instances.Values)
        {
            instance.Shutdown();
        }

        Instances.Clear();
    }

    public static bool Has(string? name = null)
    {
        if (name != null)
        {
            return Instances.ContainsKey(name);
        }

        return _defaultInstance != null;
    }
}

File: kotlin/src/main/kotlin/com/royaltics/errortracker/Tracker.kt
package com.royaltics.errortracker

import com.royaltics.errortracker.types.ClientConfig
import com.royaltics.errortracker.types.EventLevel
import java.util.concurrent.ConcurrentHashMap

object Tracker {
    private val instances = ConcurrentHashMap<String, ErrorTrackerClient>()
    private var defaultInstance: ErrorTrackerClient? = null

    fun create(config: ClientConfig, name: String? = null): ErrorTrackerClient {
        val client = ErrorTrackerClient(config)

        if (name != null) {
            instances[name] = client
        } else if (defaultInstance == null) {
            defaultInstance = client
        }

        return client.start()
    }

    fun get(name: String? = null): ErrorTrackerClient {
        return if (name != null) {
            instances[name] ?: throw IllegalStateException("Tracker instance \"$name\" not found")
        } else {
            defaultInstance ?: throw IllegalStateException("No default tracker initialized. Call Tracker.create() first.")
        }
    }

    fun error(
        error: Throwable,
        level: EventLevel = EventLevel.ERROR,
        metadata: Map<String, String>? = null
    ): ErrorTrackerClient {
        return get().error(error, level, metadata)
    }

    fun event(
        title: String,
        level: EventLevel = EventLevel.INFO,
        metadata: Map<String, String>? = null
    ): ErrorTrackerClient {
        return get().event(title, level, metadata)
    }

    suspend fun flush() {
        get().forceFlush()
    }

    fun pause(): ErrorTrackerClient {
        return get().pause()
    }

    fun resume(): ErrorTrackerClient {
        return get().resume()
    }

    fun shutdown() {
        defaultInstance?.shutdown()
        instances.values.forEach { it.shutdown() }
        instances.clear()
        defaultInstance = null
    }

    fun has(name: String? = null): Boolean {
        return if (name != null) {
            instances.containsKey(name)
        } else {
            defaultInstance != null
        }
    }
}
